Add DRS Endpoint

david4096 opened the pull request:
Adds features based on the DRS lambda first programmed at https://github.com/DataBiosphere/dos-azul-lambda .

Will address #638 

* A new path at `/ga4gh/dos/v1/dataobjects/{data_object_id}`.
* Made the service a python module so its functions can be unit tested. (2 additional `__init__.py`)
* Tests demonstrating the endpoint usage
* Notebook demonstrating using the DRS endpoint to download a file

hannes-ucsc commented:
Oh, since this is an externally observable change, an entry to the change log needs to be made.

hannes-ucsc commented on test/service/test_drs.py:
```diff
@@ -0,0 +1,99 @@
+        response.raise_for_status()
+        response_json = response.json()
+        file_id = response_json['hits'][0]['files'][0]['uuid']
+        get_url = "{}/ga4gh/dos/v1/dataobjects/{}".format(self.base_url, file_id)
```
We use f-strings.

hannes-ucsc commented on test/service/test_drs.py:
```diff
@@ -0,0 +1,99 @@
+        """
+        file_id = "NOT_A_GOOD_IDEA"
+        get_url = "{}/ga4gh/dos/v1/dataobjects/{}".format(self.base_url, file_id)
+        # Should cause a 404 error
```
Then it should assert that. Also, `with self.assertRaises(): …; self.raise_for_status()` seems convoluted. Why not simply `assertEquals(404, response.status)` ?

hannes-ucsc replied:
You must have accidentally clicked Resolve conversation. This hasn't been addressed.

hannes-ucsc commented on test/service/test_drs.py:
```diff
@@ -0,0 +1,99 @@
+        """
+        data_object, _ = self.get_data_object()
+        url = data_object['urls'][0]['url']
+        self.assertIn('http', url, "Make sure it is url-like")
```
Any string containing the string "http" is URL-like? This seems too lenient and redundant at the same time, given the fact that `requests` would barf if it weren't a URL.

hannes-ucsc commented on test/service/test_drs.py:
```diff
@@ -0,0 +1,99 @@
+            drs_response = requests.get(get_url)
+            drs_response.raise_for_status()
+
+    @unittest.skip("DSS and Azul indices do not match in testing"
```
Reading this, I get the feeling that the tests are actually hitting the DSS. That would disqualify them as unit tests, obviously. If I am correct and if at all possible, try mocking the DSS responses. test_dss_proxy.py does that.

david4096 commented:
Thanks for the comments @hannes-ucsc. I'll remove the format strings. Regarding mocking the DSS response, I believe it is possible but am OK with removing/skipping that test. Is the "fetch" surface tested elsewhere?

Some of the test patterns are copypasta from other tests there and could be less redundant.

I'll also take another look at the import patterns for unit testing so I can match them.

hannes-ucsc commented:
> Is the "fetch" surface tested elsewhere?

`/fetch/dss/files` is covered by `test_dss_proxy.py` if that's what you mean.

There is not much point in dragging around a test that we don't intend to ever unskip. If this test doesn't add any additional coverage, let's remove it. This is your call, since I haven't sufficiently thought through the cases (and shouldn't have to).

One more thing, I will get an email when you request a review on Github. But additionally, we also assign the PR to the reviewer (me). Only then will it show up when we filter on the reviewer during stand-up. And looking at the PRs in Github, I can tell at a glance which ones need my attention.

david4096 commented:
There's 1 failed and 1 errored.

```
======================================================================
ERROR: test_azul_to_drs (service.test_drs.DataRepositoryServiceEndpointTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/DataBiosphere/azul/test/service/test_drs.py", line 73, in test_azul_to_drs
    data_object = azul_to_obj(mock_azul_file)
  File "/home/travis/build/DataBiosphere/azul/lambdas/service/app.py", line 1271, in azul_to_obj
    urls = [{"url": file_url(result['uuid'], version=result['version'], replica=replica)} for replica in replicas]
  File "/home/travis/build/DataBiosphere/azul/lambdas/service/app.py", line 1271, in <listcomp>
    urls = [{"url": file_url(result['uuid'], version=result['version'], replica=replica)} for replica in replicas]
  File "/home/travis/build/DataBiosphere/azul/lambdas/service/app.py", line 811, in file_url
    url = self_url(endpoint_path=f'{proxy_endpoint_path}/{uuid}')
  File "/home/travis/build/DataBiosphere/azul/lambdas/service/app.py", line 822, in self_url
    retry_url = f'{protocol}://{base_url}{endpoint_path}'
UnboundLocalError: local variable 'protocol' referenced before assignment

======================================================================
FAIL: test_data_object_not_found (service.test_drs.DataRepositoryServiceEndpointTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/DataBiosphere/azul/test/service/test_drs.py", line 50, in test_data_object_not_found
    drs_response = requests.get(get_url)
AssertionError: HTTPError not raised

----------------------------------------------------------------------
Ran 140 tests in 762.751s

FAILED (failures=1, errors=1)
```

The reason the first is erroring is because I removed the "request mocking" for when in a test environment I had added here. https://github.com/DataBiosphere/azul/pull/642/commits/48952c849c3730fc482cfe2cc9aebf15b3877258 .

The second is an actual failure that I'll have to diagnose, and which I haven't yet replicated locally.

hannes-ucsc commented on test/service/test_drs.py:
```diff
@@ -0,0 +1,59 @@
+
+from service import WebServiceTestCase
+
+#from lambdas.service.app import azul_to_obj
```
Remove, please.

hannes-ucsc commented on lambdas/service/app.py:
```diff
@@ -780,7 +780,7 @@ def file_url(uuid, **params):
 def self_url(endpoint_path=None):
     protocol = app.current_request.headers.get('x-forwarded-proto', 'http')
-    base_url = app.current_request.headers['host']
+    base_url = app.current_request.headers.get('host')
```
We didn't need this until now. Furthermore, if the `host` header were missing, I would want to learn that here rather than …

hannes-ucsc commented on lambdas/service/app.py:
```diff
@@ -780,7 +780,7 @@ def file_url(uuid, **params):
+    base_url = app.current_request.headers.get('host')
     if endpoint_path is None:
         endpoint_path = app.current_request.context['path']
     retry_url = f'{protocol}://{base_url}{endpoint_path}'
```
… produce a URL that reads `http://None/…` and causes more insidious failures later.

hannes-ucsc commented:
Remember to assign the PR to me in addition to requesting review from me.

Also: https://github.com/DataBiosphere/azul/pull/642#issuecomment-455395109

hannes-ucsc commented on lambdas/service/app.py:
```diff
@@ -780,7 +780,7 @@ def file_url(uuid, **params):
 def self_url(endpoint_path=None):
     protocol = app.current_request.headers.get('x-forwarded-proto', 'http')
-    base_url = app.current_request.headers['host']
+    base_url = app.current_request.headers.get['host']
```
We didn't need this until now. Furthermore, if the host header were missing, I would want to learn that here rather than …

hannes-ucsc commented on CHANGELOG.yaml:
```diff
@@ -52,6 +52,17 @@ changes:
+  - title: Add GA4GH Data Repository Service GET Method
+    issues:
+      - https://github.com/DataBiosphere/azul/issues/638
+      - https://github.com/DataBiosphere/azul/pull/642
```
We don't refer to PRs here, only issues.
